api/app.py

```diff
@@ -17,69 +18,169 @@
+
+    recipe = Recipe(name=recipe_data['name'],
+                    duration=recipe_data['duration'],
+                    pictures=pictures_concat,
+                    instructions=recipe_data['instructions'],
+                    categories=category_obj)
+    db.session.add(recipe)
+    db.session.commit()
+
+    for ingredient_data in recipe_data['ingredients']:
+        ingredient = Ingredient(name=ingredient_data['name'],
+                                unit=ingredient_data['unit'],
+                                quantity=ingredient_data['quantity'],
+                                recipe_id=recipe.id)
+        db.session.add(ingredient)
+
+    db.session.commit()
+
+    return jsonify(recipe.as_dict()), 201
+
 
 @app.route('/api/recipes/<int:recipe_id>', methods=['DELETE'])
 def delete_recipe(recipe_id):
-    for recipe in recipes:
-        if recipe_id == recipe['id']:
-            recipes.remove(recipe)
-            return jsonify("adios")
-    return jsonify({"error": "404 not found"}), 404
+    recipe = db.session.get(Recipe, recipe_id)
+    if not recipe:
+        return jsonify({'error': 'Recipe not found'}), 404
+
+    Ingredient.query.filter_by(recipe_id=recipe.id).delete()
+
+    db.session.delete(recipe)
+    db.session.commit()
+    return jsonify({'message': f'Recipe with ID {recipe_id} deleted successfully'}), 204
+
 
 @app.route('/api/recipes/<int:recipe_id>', methods=['PUT'])
 def edit_recipe(recipe_id):
-    for recipe in recipes:
-        if recipe_id == recipe['id']:
-            recipe['name'] = categories if (categories := request.json.get('name')) else recipe['name']
-            recipe['duration'] = categories if (categories := request.json.get('duration')) else recipe['duration']
-            recipe['pictures'] = categories if (categories := request.json.get('pictures')) else recipe['pictures']
-            recipe['instructions'] = categories if (categories := request.json.get('instructions')) else \
-                recipe['instructions']
-            recipe['categories'] = categories if (categories := request.json.get('categories')) else \
-                recipe['categories']
-            recipe['ingredients'] = categories if (categories := request.json.get('ingredients')) else \
-                recipe['ingredients']
-            return recipe
-    return jsonify({"error": "404 not found"}), 404
+    recipe = db.session.get(Recipe, recipe_id)
+    if not recipe:
+        return jsonify({'error': 'Recipe not found'}), 404
+
+    data = request.json
+
+    if data['name']:
+        recipe.name = data['name']
+    if data['duration']:
+        recipe.duration = data['duration']
+    if data['pictures']:
+        recipe.pictures = ",".join(data['pictures'])
+    if data['instructions']:
+        recipe.instructions = data['instructions']
+
+    if data['categories']:
+        recipe.categories.clear()
+
+        for cat_data in data['categories']:
+            category = Category.query.get(cat_data["id"])
```

Comment:
Same as in `create_recipe()`, we receive the `categories` as a list of dicts from frontend, but they don't have the `id` field.